drivers/flash/spi_nor.c

```diff
@@ -48,36 +48,58 @@ LOG_MODULE_REGISTER(spi_nor, CONFIG_FLASH_LOG_LEVEL);
 
 #define SPI_NOR_MAX_ADDR_WIDTH 4
 
-#if DT_INST_NODE_HAS_PROP(0, t_enter_dpd)
-#define T_DP_MS DIV_ROUND_UP(DT_INST_PROP(0, t_enter_dpd), NSEC_PER_MSEC)
-#else /* T_ENTER_DPD */
-#define T_DP_MS 0
-#endif /* T_ENTER_DPD */
-#if DT_INST_NODE_HAS_PROP(0, t_exit_dpd)
-#define T_RES1_MS DIV_ROUND_UP(DT_INST_PROP(0, t_exit_dpd), NSEC_PER_MSEC)
-#endif /* T_EXIT_DPD */
-#if DT_INST_NODE_HAS_PROP(0, dpd_wakeup_sequence)
-#define T_DPDD_MS DIV_ROUND_UP(DT_INST_PROP_BY_IDX(0, dpd_wakeup_sequence, 0), NSEC_PER_MSEC)
-#define T_CRDP_MS DIV_ROUND_UP(DT_INST_PROP_BY_IDX(0, dpd_wakeup_sequence, 1), NSEC_PER_MSEC)
-#define T_RDP_MS DIV_ROUND_UP(DT_INST_PROP_BY_IDX(0, dpd_wakeup_sequence, 2), NSEC_PER_MSEC)
-#else /* DPD_WAKEUP_SEQUENCE */
-#define T_DPDD_MS 0
-#endif /* DPD_WAKEUP_SEQUENCE */
+#define ANY_INST_HAS_MXICY_MX25R_POWER_MODE DT_ANY_INST_HAS_PROP_STATUS_OKAY(mxicy_mx25r_power_mode)
+
+/* boolean prop, don't use DT_INST_HAS_PROP */
+#define INST_HAS_DPD_OR(idx) (DT_INST_PROP(idx, has_dpd)) ||
+#define ANY_INST_HAS_DPD DT_INST_FOREACH_STATUS_OKAY(INST_HAS_DPD_OR) 0
```

Comment:
using DT_INST_NODE_HAS_PROP_AND_OR doesn't seem to work
it expands to : 0 || and breaks the build due to right operand missing 
also, has_dpd is a boolean prop so I think I shouldn't use HAS_PROP constructs but DT_PROP
https://docs.zephyrproject.org/latest/build/dts/api/api.html#group__devicetree-generic-prop_1ga8e1fd9ebacd85d2013df027d041d506b
`This tests whether the property is defined at all, not whether a boolean property is true or false. To get a boolean property’s truth value, use [DT_PROP(node_id, prop)] instead `

I don't see how to combine DT_INST_FOREACH_STATUS_OKAY with UTIL_OR
can you show me an example please?

Reply:
put a 0 on the end

Reply:
So using this construct 
`#define INST_HAS_DPD_OR(idx) DT_INST_NODE_HAS_PROP_AND_OR(idx, has_dpd)`
`#define ANY_INST_HAS_DPD DT_INST_FOREACH_STATUS_OKAY(INST_HAS_DPD_OR) 0`

the code compiles but the behavior is not completely correct.
Specifically the following struct elements:
```
#if ANY_INST_HAS_DPD
	uint16_t t_enter_dpd; /* in microseconds */
	uint16_t t_dpdd_ms;   /* in microseconds */
	uint16_t t_exit_dpd;  /* in microseconds */
#endif

```
With an overlay where has-dpd is true
`...
has-dpd;
...`
the field .dpd_exist is initialized to 1 correctly.
and these fields are defined at the struct level correctly
```
 uint16_t t_enter_dpd;
 uint16_t t_dpdd_ms;
 uint16_t t_exit_dpd;
```

But with an overlay where has-dpd is false
`...
/delete-property/ has-dpd;
...`
the field .dpd_exist is initialized to 0 correctly.
But these fields are defined at the struct level
```
 uint16_t t_enter_dpd;
 uint16_t t_dpdd_ms;
 uint16_t t_exit_dpd;
```

On a functional level this is not a problem, because all usage of those fields are inside conditionals
`	if (driver_config->dpd_exist) {..}`

But it's a waste of memory in the case where all used devices do not have dpd

This behavior is avoided with the original solution
`#define INST_HAS_DPD_OR(idx) (DT_INST_PROP(idx, has_dpd)) ||
#define ANY_INST_HAS_DPD DT_INST_FOREACH_STATUS_OKAY(INST_HAS_DPD_OR) 0`

Which one is better?

Reply:
OK, sorry you are right the DT_INST_FOREACH_STATUS_OKAY is hard to use here.
I have proposed PR to your PR: https://github.com/ashiroji/zephyr/pull/1 that provides improved macros, you can try them and check if they suit you.

I have also proposed change to the DT_ANY_INST_HAS_PROP_STATUS_OKAY  https://github.com/zephyrproject-rtos/zephyr/pull/70385, but it only fixes half of what you need.

Reply:
@de-nordic , Sorry, I seem to have made a mistake when updating the PR.
I'll try to fix it.